### ISSUE TITLE
[css-zoom?] Zoom and inheritance. 

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#9397]
Zoom has some interesting behavior that makes no sense to me ([live](https://www.software.hixie.ch/utilities/js/live-dom-viewer/saved/12028)):

```
<!doctype html>
<div style="font-size: 16px; width: 100px; line-height: 16px;">
  <div style="background-color: green; font-size: inherit; line-height: inherit; width: inherit; height: 100px; zoom: 2;">What size am I?</div>
</div>
```

That for some reason ends up with:

 * `line-height` of `16px`
 * `font-size` of `32px`
 * `width` of `100px`
 * `height` of `200px`

That just makes no sense. This happens because the way `zoom` is implemented in browsers is by mutating the computed value, so inheritance doesn't account for zoom since it copies the un-zoomed computed value.

But then somehow `font-size` (but not `line-height`) is special, and gets recomputed on zoom changes??

cc: #5623 @chrishtr @lilles @smfr @tabatkins @atanassov

**Comment by SelenIT** [comment]
Chrome 116 (M1 Mac) reports the following computed styles of this element:
<img width="368" alt="Screenshot 2023-09-25 at 00 24 35" src="https://github.com/w3c/csswg-drafts/assets/282754/3b97e7a0-1793-4cce-883a-8de9d85443ee">
It seems that font-size is unchanged, but all pixel values related to the parent (width and inherited line-height) get divided by the zoom factor. Since the "parent pixel" equals to half of the "zoomed pixel" of the element, treating all such values consistently makes some sense to me.

**Comment by lilles** [comment]
To confirm this is not line-height specific I tried inheriting border-left-width explicitly, which also ends up with not applying zoom in Chrome.

**Comment by lilles** [comment]
Implicit inheritance of line-height also doesn't work.

**Comment by lilles** [comment]
And getComputedStyle() un-zooms based on the effective zoom which means it returns 8px for the inherited line-height/width etc. in the case above.

I'm not too concerned about fixing this for explicit inheritance of non-inherited properties in Blink, but adjusting inherited properties when the effective zoom changes will probably be expensive. Hopefully, changing the zoom property won't happen on many elements.

**Comment by emilio** [comment]
Yeah, my preference would be not to adjust on _any_ inheritance (so, removing the font-size special-case).

**Comment by chrishtr** [comment]
I don't think we have evidence that changes of zoom are common or performance-sensitive, so I am not worried about performance. I don't know of any use cases for zoom that need a super fast relayout on change of zoom. On the contrary, the use cases I know of are all about getting responsive design right, which I think is more important than performance in this case.

I am, however, worried about web compat and usefulness of this API. From that perspective, applying zoom to all inherited lengths (and not just font size) that are not percentages is consistent and makes the most sense.

**Comment by emilio** [comment]
Note that "changes of zoom" here doesn't mean just dynamic changes. Any change of zoom from ancestor to descendant would incur that extra performance cost. So `* { zoom: 101%; }` would cause a _lot_ of extra work (because each element would zoom a little bit more than the previous etc), and there doesn't need to be any dynamic change for that.

**Comment by chrishtr** [comment]
> So `* { zoom: 101%; }` would cause a _lot_ of extra work (because each element would zoom a little bit more than the previous etc), and there doesn't need to be any dynamic change for that.

A lot of extra work due to processing inherited lengths? I don't get how, could you explain? In that situation it seems you'd have to multiply all the lengths by some numbers regardless.

(also, `* { zoom: 101%; }` seems as bad of a footgun as `* { font-size: 101%; }`, why would developers do that?)

**Comment by emilio** [comment]
Right, so the work is the same as increasing the font-size, but times the amount of inherited length properties, right? Given engines rely on heavy sharing of inherited computed style data in order to have somewhat reasonable memory usage, it seems problematic, though I haven't checked by hand the number of affected properties.

**Comment by emilio** [comment]
But yeah I guess changing zoom to something != 1 might be uncommon enough in practice. Still I've seen pages with hundreds of individual icons zoomed in/out.

**Comment by emilio** [comment]
Another thing I haven't thought in depth about is custom properties. Presumably such adjustment would have to be done for typed inherited custom properties, but not for untyped ones (which are more common). I think that would be rather confusing for authors tho...

**Comment by emilio** [comment]
> Yes. But is that big deal? Also, can't the browsers optimize it by computing it once and passing along a cached value to descendants that also inherit?

Not quite sure I follow. Consider the case of a bunch of leaf elements (images) that have a zoom value. I've [seen this in the wild](https://bugzilla.mozilla.org/show_bug.cgi?id=1861726).

The problem has nothing to do with descendants, but for each of those images now you need to do a significantly larger amount of work.

> Wouldn't sharing still be possible, because descendants that don't change font-size or other lengths can reuse each others' values through caching?

The problem is not inheriting to further descendants IMO, it's generally breaking the parent-ancestor sharing at zoom boundaries too. Again in the case above, none of those images would share memory for the relevant data structures if we did this.
